src/v/cluster/rm_stm.cc

```diff
@@ -1914,6 +1943,20 @@ model::offset rm_stm::last_stable_offset() {
         return model::invalid_lso;
     }
 
+    auto use_applied = true;
+    auto synced_leader = _c->is_leader() && _c->term() == _insync_term;
+    synced_leader = synced_leader && _mem_state.term == _insync_term;
+    if (synced_leader) {
+        // if there are inflight eos reqs we can't guess the right LSO
+        // because the outcome of the inflight requests may affect it
+        use_applied = _mem_state.inflight_eos != 0;
+        // even when a request is replicated we still need to wait until
+        // an STM catches up with it to make sure LSO reflects it
+        use_applied = use_applied || _mem_state.last_eos > last_applied;
```

Comment:
same with this.. if there is any batch flushed after last_visible, we still fallback to applied offset? That seems to strict.. don't we just need to consider non_applied txns in [last_applied, hwm] range?

Reply:
fixed